source/simulator/core.cc

```diff
@@ -1767,15 +1767,53 @@ namespace aspect
     // start the principal loop over time steps:
     do
       {
-        // Only solve if we are not in pre-refinement, or we do not want to skip
-        // solving in pre-refinement.
+        // During pre-refinement, do not solve if we are asked to skip it:
         if (! (parameters.skip_solvers_on_initial_refinement
                && pre_refinement_step < parameters.initial_adaptive_refinement))
           {
             start_timestep ();
 
             // then do the core work: assemble systems and solve
-            solve_timestep ();
```

Comment:
I somewhat dislike adding so many lines in the main run loop that are unrelated to it. Conceptionally it is the responsibility of the solve_timestep function to handle non-convergence. Could we move the whole try-catch block into solve_timestep, and modify the code in `run` to the following:
```
const bool retry_timestep = solve_timestep();
if (retry_timestep)
  continue;
```

All of the exception handling can stay the same, but it keep this loop readable.

Reply:
please check my comment from 2019: https://github.com/geodynamics/aspect/pull/2744/files#r244753861
does this still apply?